### DISCUSSION TITLE
Python based Admin Interface Comparison

### DISCUSSION THREAD

**Discussion by hasansezertasan** [jowilf/starlette-admin#333] [Show and tell]
# Python based Admin Interface Comparison

Hi everyone :wave:.

I'm planning to improve "Alternatives, Inspiration and Comparisons" section in the docs. I want to add a comparison table with all the features. I used Flask Admin and SQLAdmin for some time but nearly starting to use Starlette Admin. I'd like to know what are the main differences, which one has a specific feature and which one doesn't? Are there any other alternatives that is not listed below?

## Feature Parity and Comparison Table

| Feature              | [Django Admin]     | [Flask Admin]      | [Flask App Builder] | [Piccolo Admin]    | [FastAPI Admin]    | [FastAdmin]        | [SQLAdmin]         | [FastAPI Amis Admin] | [ohmyadmin]        | [fastapi-react-admin] | [Starlette Admin]  |
| -------------------- | ------------------ | ------------------ | ------------------- | ------------------ | ------------------ | ------------------ | ------------------ | -------------------- | ------------------ | --------------------- | ------------------ |
| Frontend Stack       | Template Engine    | Template Engine    | Template Engine     | VueJS              | Template Engine    | ReactJS            | Template Engine    | Amis                 | Mixed              | ReactJS               | Template Engine    |
| Custom Views         | :white_check_mark: | :white_check_mark: | :white_check_mark:  | :white_check_mark: | :white_check_mark: | :question:         | :white_check_mark: | :white_check_mark:   | :question:         | :question:            | :white_check_mark: |
| CRUDL                | :white_check_mark: | :white_check_mark: | :white_check_mark:  | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark:   | :question:         | :question:            | :white_check_mark: |
| Search               | :white_check_mark: | :white_check_mark: | :question:          | :white_check_mark: | :question:         | :question:         | :question:         | :question:           | :question:         | :question:            | :white_check_mark: |
| Authentication       | :white_check_mark: | :white_check_mark: | :white_check_mark:  | :white_check_mark: | :question:         | :question:         | :white_check_mark: | :question:           | :question:         | :question:            | :white_check_mark: |
| Batch Actions        | :white_check_mark: | :white_check_mark: | :question:          | :white_check_mark: | :question:         | :question:         | :white_check_mark: | :question:           | :question:         | :question:            | :white_check_mark: |
| Row Actions          | :white_check_mark: | :white_check_mark: | :question:          | :question:         | :question:         | :question:         | :question:         | :question:           | :question:         | :question:            | :white_check_mark: |
| Internationalization | :white_check_mark: | :white_check_mark: | :white_check_mark:  | :white_check_mark: | :white_check_mark: | :question:         | :question:         | :white_check_mark:   | :question:         | :question:            | :white_check_mark: |
| [SQLAlchemy]         | :x:                | :white_check_mark: | :white_check_mark:  | :x:                | :x:                | :white_check_mark: | :white_check_mark: | :white_check_mark:   | :white_check_mark: | :white_check_mark:    | :white_check_mark: |
| [MongoEngine]        | :question:         | :white_check_mark: | :warning:           | :x:                | :x:                | :x:                | :x:                | :x:                  | :x:                | :x:                   | :white_check_mark: |
| ...                  | :question:         | :question:         | :question:          | :question:         | :question:         | :question:         | :question:         | :question:           | :question:         | :question:            | :question:         |

Emoji Key:

| Emoji              | Meaning |
| ------------------ | ------- |
| :white_check_mark: | Yes     |
| :x:                | No      |
| :warning:          | Partial |
| :question:         | Unknown |

## Other Tools

- [fa.bootstrap] - "pyramid_formalchemy twitter bootstrap support"

I'm planning to update this table in the future but I won't include some features that are exceptional like how Flask Admin has a console for redis, File and Folder manager and how FastAPI Amis Admin has some sort of UI/Theme editor...

Don't forget to recommend more...

[Django Admin]: https://docs.djangoproject.com/en/4.2/ref/contrib/admin/
[Flask Admin]: https://github.com/flask-admin/flask-admin
[Flask App Builder]: https://github.com/dpgaspar/Flask-AppBuilder
[Piccolo Admin]:https://github.com/piccolo-orm/piccolo_admin
[FastAPI Admin]:https://github.com/fastapi-admin/fastapi-admin
[FastAdmin]:https://github.com/vsdudakov/fastadmin
[FastAPI Amis Admin]:https://github.com/amisadmin/fastapi-amis-admin
[SQLAdmin]:https://github.com/aminalaee/sqladmin
[ohmyadmin]: https://github.com/alex-oleshkevich/ohmyadmin
[Starlette Admin]:https://github.com/jowilf/starlette-admin
[SQLAlchemy]: https://github.com/sqlalchemy/sqlalchemy
[MongoEngine]: https://github.com/MongoEngine/mongoengine
[fastapi-react-admin]: https://github.com/enveloss/fastapi-react-admin
[fa.bootstrap]: https://github.com/RedTurtle/fa.bootstrap

**Comment by agxc** [comment]
Do you intend to compare with flask-appbuilder?  It is widely used by big organisation like apache org and Airbnb for flask related products. It basically supports all the features here plus auto restapi generation, security setting and many more.

e.g.
https://flask-appbuilder.readthedocs.io/en/latest/rest_api.html#model-rest-api

**Reply by hasansezertasan** [reply]
Yes, Flask App Builder is already in the table. I didn't know that apache.org or airbnb used Flask App Builder, or Flask... This is surprising. Auto RestAPI generation looks great but in general, utilities and out-of-the-box functions looks quite well over there, especially permissions. These alternatives could also inspire us to improve starlette-admin.

**Reply by agxc** [reply]
Yeah, I just found out that Apache Superset, a top-tier product from Apache, is actually built heavily on Flask-Appbuilder (FAB). The author who made FAB is on the Apache Superset committee. This is probably why some big Apache projects like Airflow is also built using FAB! Recommend you look into their Github repos for your interest. 

I dove deep into FAB source code, and honestly, it's probably one of the most decked-out packages I've come across. The code? 🔥 Top-notch and has been put to the test in some of the most advanced products out there.

I was thinking of using FAB for a project at work, but the default permission/auth model threw a wrench in my plans. It's a bit too rigid for tweaking to fit my org's needs, so had to pass on it.

One thing leads to another, I found out the starlette-admin recently. You know what, I'm a big fan now. Feels like it carries the same assumption-free philosophy as flask-admin. Has tons of potential to be the best one. 

I would advise against having a built-in security model .

**Reply by hasansezertasan** [reply]
OMG, I didn't know that Flask App Builder were this popular. I just checked their repo, it looks quite interesting. I believe we can get inspired by this.

I'm working on a security module. I have implemented security layers for Flask Admin, Starlette, Starlette Admin and FastAPI. Its not there yet but one day, I hope I can publish it.

I also have opened a PR at #372 to include more alternatives and add reference to this discussion. I believe this module will shine in the future and more people will be attracted by its power. Did you know this project is one of the requirements of [reflex](https://reflex.dev/) project.

**Reply by hasansezertasan** [reply]
I have also found this question: [python - Flask-Admin vs Flask-AppBuilder - Stack Overflow](https://stackoverflow.com/questions/30126607/flask-admin-vs-flask-appbuilder)

Did you now about [valhuber/ApiLogicServer](https://github.com/valhuber/ApiLogicServer#readme)?

**Comment by agxc** [comment]
Hi @hasansezertasan , this is a fascinating discussion. I appreciate you sharing these resources. I've come across API Logic Server but I've only had a closely look after you've mentioned here. It looks pretty amazing. Also, it integrates Flask Appbuilder too based on the documentation.
 
https://valhuber.github.io/ApiLogicServer/Working-with-Flask-AppBuilder/

I've also read the comparison thread between flask-admin and F.A.B, but that was back in 2015. The FAB has evolved so much since ethen and it is now used as the core extension for some of the biggest Flask  based projects out there. 

Overall, I feel Flask has a quite mature ecosystem. I have been struggling to find good/mature extensions for Starlette/FastAPI However, starlette-admin is probably the most modern admin panel out there and a lot faster than flask based admin panels.

**Comment by hasansezertasan** [comment]
I have found other packages:

- [mixartemev/femto-admin: Simplest fastest minimal fully native async auto generating zero config one line app ASGI CRUD Admin panel for Tortoise ORM models](https://github.com/mixartemev/femto-admin)
- [encode/dashboard: An admin interface for ASGI Web frameworks.](https://github.com/encode/dashboard)

**Reply by sinisaos** [reply]
@hasansezertasan Can you please update your comparison table for Piccolo Admin as [authentication](https://piccolo-admin.readthedocs.io/en/latest/authentication/index.html), [search (filtering)](https://piccolo-admin.readthedocs.io/en/latest/index.html), [bulk actions](https://piccolo-admin.readthedocs.io/en/latest/actions/index.html) and [internationalization](https://piccolo-admin.readthedocs.io/en/latest/internationalization/index.html) are supported out of the box. There are many more features such as [custom forms](https://piccolo-admin.readthedocs.io/en/latest/custom_forms/index.html) (from Pydantic models), [single and multiple file uploads](https://piccolo-admin.readthedocs.io/en/latest/media_storage/index.html), a [rich text editor](https://piccolo-admin.readthedocs.io/en/latest/table_config/index.html#rich-text-columns), and various configurations via [TableConfig](https://piccolo-admin.readthedocs.io/en/latest/table_config/index.html). Thanks in advance.

**Reply by hasansezertasan** [reply]
Thank you @sinisaos for enlightening me 🙏. I just updated the table. I plan to make a more comprehensive comparison and create a dedicated repository. I'll definitely put some introduction about custom forms on it 🚀.

If you know any other python-based admin interface projects like these that are not listed on the table, please let me know 😇.

**Reply by sinisaos** [reply]
No problem. Thanks for your update. I think you should also update Django Admin because Django has all the features you listed in the table except it doesn't support SQLAlchemy. I don't have much experience with SQLAlchemy (although it's actually the main ORM outside of Django), but Starlette Admin is the best option for me outside of Django and Piccolo (I even made a small contribution to Starlette Admin a while ago)

**Reply by hasansezertasan** [reply]
You're welcome. I don't have much experience with Django so I didn't know that but thanks. I'll update the Django column with ✅s. One question, does Django Admin support MongoEngine?

Starlette Admin is also the best option for me currently but I couldn't make peace with Piccolo, maybe it's just because I'm more of an SQLAlchemy person, or maybe because it's written with VueJS. Not because I have a problem with VueJS, I just want to be in the game, easily customize and override everything, in Python (`.py`) 🐍 environment. VueJS prevents me from doing so but not Jinja...

> I even made a small contribution to Starlette Admin a while ago.

I believe I do remember that. You added `TinyMCEEditorField` at #131. I believe you might be interested in #499.

**Reply by sinisaos** [reply]
> You're welcome. I don't have much experience with Django so I didn't know that but thanks. I'll update the Django column with ✅s. One question, does Django Admin support MongoEngine?
> 

I don't know I never use Django with MongoDB

> Starlette Admin is also the best option for me currently but I couldn't make peace with Piccolo, maybe it's just because I'm more of an SQLAlchemy person, or maybe because it's written with VueJS. 

In my case it is quite opposite. I know it's not very popular, but I like Django ORM and Piccolo ORM a lot more than SQLAlchemy. I'm not an expert in SQLAchemy but somethimes SQLAlchemy has complicated solutions, the documentation is not easy to search, I haven't found an easy way for migrations per app like in Django or Piccolo etc. But that's only personal preferences.

> Not because I have a problem with VueJS, I just want to be in the game, easily customize and override everything, in Python (`.py`) 🐍 environment. VueJS prevents me from doing so but not Jinja...

Piccolo Admin has a separate backend and frontend and that's nice, but you're right that it's much harder to customize things than just in Jinja (Python), because everything has to be done in both Python and Vue.

**Reply by hasansezertasan** [reply]
> I don't know I never use Django with MongoDB

Alright, that stays as ❓ then.

> In my case it is quite opposite. I know it's not very popular, but I like Django ORM and Piccolo ORM a lot more than SQLAlchemy. I'm not an expert in SQLAchemy but somethimes SQLAlchemy has complicated solutions, the documentation is not easy to search, I haven't found an easy way for migrations per app like in Django or Piccolo etc. But that's only personal preferences.

I believe Django ORM is quite popular. I hear good things about it but as far as I know, it's limited to Django, so that's a problem to me... You are right in your complaints about SQLAlchemy 🥇.

> Piccolo Admin has a separate backend and frontend and that's nice, but you're right that it's much harder to customize things than just in Jinja (Python), because everything has to be done in both Python and Vue.

That's right. Decoupling backend and frontend is quite nice... But lately, I started to think otherwise. I'm highly exposed to terms like HATEOAS, HDA (Hypermedia Driven Applications, which Django adopts this approach), and SSR (Server Side Rendering). They are quite sexy! So I like to couple frontend and backend at the moment (even though I don't like template engines like Mako and Jinja2 because of the micro-syntax) and no, not BFF (Backend for Frontend) 😮‍💨.

**Comment by asmaier** [comment]
There is also https://benavlabs.github.io/crudadmin/ .

**Comment by vsdudakov** [comment]
Late addition, since the thread's still collecting these: FastAdmin
(https://github.com/vsdudakov/fastadmin) — I'm the author, so weigh the bias
accordingly.

Against the table's axes: Django-Admin-style API (`list_display`, `list_filter`,
`fieldsets`, inlines, actions), and its main difference from the others listed is
that it's multi-framework *and* multi-ORM in one package — mounts into FastAPI,
Flask, or Django, on top of Tortoise, SQLAlchemy (async), Django ORM, or Pony.
React/antd frontend ships bundled, so no Node at install time. MIT, fully typed.

Happy to fill in a full row of the comparison table if that's still being maintained.
